sygnal/helper/proxy/proxy_asyncio.py

```diff
@@ -0,0 +1,241 @@
+        # method on it to open a tunnelled TCP connection through the proxy to
+        # the other side
+        transport.write(f"CONNECT {self.target_hostport} HTTP/1.1\r\n".encode())
+        parts = self.proxy_url_parts
+        transport.write(f"Host: {parts.hostname}:{parts.port or 80}\r\n".encode())
+        if parts.username is not None and parts.password is not None:
+            # a credential pair is a urlsafe-base64-encoded pair separated by colon
+            encoded_credentials = urlsafe_b64encode(
+                f"{parts.username}:{parts.password}".encode()
+            )
+            transport.write(
+                b"Proxy-Authorization: basic " + encoded_credentials + b"\r\n"
+            )
+        # a blank line terminates the request headers
+        transport.write(b"\r\n")
+
+        logger.debug("Initiating proxy CONNECT")
+
+        # now we wait ...
+        self.transport = transport
+
+
+class ProxyingEventLoopWrapper:
+    """
+    This is a wrapper for an asyncio.AbstractEventLoop which intercepts calls to
+    create_connection and transparently tunnels them through an HTTP CONNECT
+    proxy.
+    """
+
+    def __init__(
+        self, wrapped_loop: asyncio.AbstractEventLoop, proxy_url_str: str,
+    ):
+        """
+        Args:
+            wrapped_loop:
+                the underlying Event Loop to wrap
+            proxy_url_str (str):
+                The address of the HTTP proxy to use.
+                Used to connect to the proxy, as well as in the `Host` request
+                header to the proxy, and for the extraction of basic
+                authentication credentials (if required).
+
+                Examples: 'http://127.0.3.200:8080'
+                       or 'http://user:secret@prox:8080'
+        """
+        self._wrapped_loop = wrapped_loop
+        self.proxy_url_str = proxy_url_str
+
+    async def create_connection(
+        self,
+        protocol_factory: Callable[[], asyncio.Protocol],
+        host: str,
+        port: int,
+        ssl: Union[bool, SSLContext] = False,
+    ):
+        proxy_url_parts = decompose_http_proxy_url(self.proxy_url_str)
+
+        def make_protocol():
+            proxy_setup_protocol = HttpConnectProtocol(
+                proxy_url_parts, f"{host}:{port}"
+            )
+            return proxy_setup_protocol
+
+        # create a raw TCP connection to the proxy
+        # (N.B. if we want to ever use TLS to the proxy [e.g. to protect the proxy
+        # credentials], we can ask this to give us a TLS connection).
+        transport, connect_protocol = await self._wrapped_loop.create_connection(
+            make_protocol, proxy_url_parts.hostname, proxy_url_parts.port
+        )
+
+        assert isinstance(connect_protocol, HttpConnectProtocol)
+
+        # wait for the HTTP Proxy CONNECT sequence to complete
+        transport = await connect_protocol.wait_for_establishment
```

Comment:
is there any reason to imagine that this `transport` could be different to the previous `transport`? if so, could you add a comment explaining? If not, I'd suggest :

```suggestion
        await connect_protocol.wait_for_establishment
```